datafusion/physical-expr/src/tree_node/mod.rs

```diff
@@ -0,0 +1,266 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! This module provides common traits for visiting or rewriting tree nodes easily.
+//!
+//! It's a duplication of the one in the crate `datafusion-common`.
+//! In the future, if the Orphan rule is relaxed for Arc<dyn T>, these duplicated codes can be removed.
```

Comment:
It would be great if we could figure out how to avoid this duplication and the one in physical-expr?

Maybe we could create a blanket impl in `datafusion-common` like

```rust
impl<T: TreeNode> TreeNode for Arc<T> {
...
}
```
?

Reply:
Thanks @alamb. However, I'm still not sure how to apply to the trait `ExecutionPlan` and `PhysicalExpr`

Reply:
I played around with this and I came up with a solution -- here is a proposed modification to this PR: https://github.com/yahoNanJing/arrow-datafusion/pull/70